lint:container command fails on named argument service

### Symfony version(s) affected

5.4.31

### Description

I receive an error like:
```
[ERROR] Invalid definition for service "service.one": argument 1 of
"ServiceOne::__construct()" accepts                    
         "Psr\Log\LoggerInterface", "null" passed.           
```

When executing
```
bin/console lint:container
```

In debugging I noticed it comes from `CheckTypeDeclarationsPass` where `::checkTypeDeclarations` expects a numbered list of arguments. I wonder why that just now fails as we have some more of these already, but anyway.

I checked the class in Symfony 6 and 7 and noticed @nicolas-grekas fixed that in PR https://github.com/symfony/symfony/pull/49944 which does sound like something different in general. Picking just the change of the for loop would fix the command for me.

### How to reproduce

Create a service with named arguments (not sure if this itself is really enough)

### Possible Solution

_No response_

### Additional Context

_No response_

--- a/src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php
+++ b/src/Symfony/Component/DependencyInjection/Compiler/CheckTypeDeclarationsPass.php
@@ -139,11 +139,17 @@ final class CheckTypeDeclarationsPass extends AbstractRecursivePass
         $envPlaceholderUniquePrefix = $this->container->getParameterBag() instanceof EnvPlaceholderParameterBag ? $this->container->getParameterBag()->getEnvPlaceholderUniquePrefix() : null;
 
         for ($i = 0; $i < $checksCount; ++$i) {
-            if (!$reflectionParameters[$i]->hasType() || $reflectionParameters[$i]->isVariadic()) {
+            $p = $reflectionParameters[$i];
+            if (!$p->hasType() || $p->isVariadic()) {
+                continue;
+            }
+            if (\array_key_exists($p->name, $values)) {
+                $i = $p->name;
+            } elseif (!\array_key_exists($i, $values)) {
                 continue;
             }
 
-            $this->checkType($checkedDefinition, $values[$i], $reflectionParameters[$i], $envPlaceholderUniquePrefix);
+            $this->checkType($checkedDefinition, $values[$i], $p, $envPlaceholderUniquePrefix);
         }
 
         if ($reflectionFunction->isVariadic() && ($lastParameter = end($reflectionParameters))->hasType()) {

